### DISCUSSION TITLE
Precognition.nvim motion as hinlay hint

### DISCUSSION THREAD

**Discussion by DrummyFloyd** [LazyVim/LazyVim#3557] [Recipes]
Because we keep learning on nvim every day x)

Seems to be a good stuff to add to ingest the good practice of vim Motion and more 

And obviously because the GA is out , I propose 
This plugin as Extra :) 

https://github.com/tris203/precognition.nvim?tab=readme-ov-file

small integration with lazyVim =) 

```lua
-- precognition.lua
local enable = false

-- Plugin configuration
-- updated becasue toggle() return a value now
-- https://github.com/tris203/precognition.nvim/pull/73
return {
  {
    "tris203/precognition.nvim",
    -- event = "LazyFile",
    opts = {
      startVisible = false,
    },
    keys = {
      {
        "<leader>uP",
        function()
          if require("precognition").toggle() then
            LazyVim.info("Precognition is on")
          else
            LazyVim.warn("Precognition is off")
          end
        end,
        desc = "toggle Precognition",
      },
    },
  },
}

```

**Comment by dpetka2001** [comment]
Probably best as a plugin that users install themselves. Doesn't provide that much in a Neovim distro in my personal opinion.

**Reply by Aman9das** [reply]
Neovim distros main aim is to improve the ease of use and adoption of neovim/vim

This plugin is critical to that purpose. 

I'll say it's a must have as training wheels extra.

**Reply by dpetka2001** [reply]
Kindly have a read at [Contributing](https://github.com/LazyVim/LazyVim/blob/8346fa7ddc9390312408be571c6443c0b57b9f21/CONTRIBUTING.md?plain=1#L15-L16). This is a rather straightforward configuration for this plugin, that could be easily adopted in the personal user configuration instead.

**Reply by AlexAllocated** [reply]
I have to agree that this would make a great addition as an extra. It's easy to add yourself yes, but the newbie is the exact demographic that won't find it so easy to add and also enable in which-key and any other enhancements that might come later on. As a newbie I'd have really appreciated being able to just enable this extra. Eventually I'd have turned it off but for a while it would have been a great help during the newbie phase where adding even one custom plugin was still a chore for me.

**Reply by dpetka2001** [reply]
```lua
-- ~/.config/nvim/lua/plugins/precognition.lua
return {
  {
    "tris203/precognition.nvim",
    opts = {},
  },
}
```
This is all it's needed for the plugin to install and load with default config. Hardly a candidate for an Extra. It's just 5 lines of code. I've just tried it like this and it works out of the box.

**Reply by AlexAllocated** [reply]
I'm aware. Hence why I mentioned integration with which-key for the toggle, which makes this infinitely more useful. It takes time to understand lazy.nvim, LazyVim, and the relationships between them all, including the plugin-specific specs. For a while it's just one big blob of Lua that takes time to mentally compartmentalize. That's why a simple addition like this would be useful since this is a plugin that could be useful even while in that newbie haze. You don't have to agree. I'm just saying this would have been nice when I personally was at that stage. Now that all the moving parts are simple in my head then yes, it's easy. But there was a short period where this was not easy:

1. I wouldn't have been aware of this useful newbie plugin because I was already overwhelmed learning LazyVim-specific things.
2. Even if I was aware, it took me a bit to start adding my own plugin specs. Even simple ones.
3. Even if I was aware of it and was at the stage of adding my own specs, your simplified implementation is exactly what I would have done, missing out on the toggle integration. I would have just added a new spec for the toggle and called it a day.

If this were an extra as @DrummyFloyd supplied it then I'd have found it quickly and enabled it as one of the first things I did was explore extras and explore which-key binds as they helped tremendously in finding the tools available in LazyVim.

PS - I'm giving reasons why I think this could be a useful exception to the contributing guidelines because of the added discoverability at a critical learning stage. You don't have to keep beating us over the head with it. If the maintainer agrees with you then great, but until then this is just a discussion.

**Reply by dpetka2001** [reply]
Just add 
```lua
keys = {
  { "key_that_you_want", "<cmd>Precognition toggle<cr>", desc = "Toggle Precognition" },
},
```
And you have a key to toggle it. 

[Contributing](https://github.com/LazyVim/LazyVim/blob/8346fa7ddc9390312408be571c6443c0b57b9f21/CONTRIBUTING.md?plain=1#L15-L16) was written by the maintainer for a reason. In my personal opinion as a simple user, users will have to put in the effort to learn how to customize their configuration according to their needs and not just expect to just press a key to make something work like in GUI editors. 

Also that's the point of the Discussions section. Newbie users can ask questions here and they will get help for what they want to do. That's not reason for such a straightforward plugin configuration to be an Extra according to `Contributing`.

**Reply by Aman9das** [reply]
`<leader>u?` is a good keymap for toggling precognition I think.

Also maybe there can be other configrations too

**Reply by dpetka2001** [reply]
`<leader>u` is already used by which-key as the group name for `+ui`. So, you better choose something else that doesn't conflict with default LazyVim keymaps.

**Reply by Aman9das** [reply]
Obviously I meant `?` after `<leader>u`.

**Reply by dpetka2001** [reply]
Oh sorry, I missed that. I thought it was a questionmark for asking a question and not part of the keymap. My bad.

**Reply by dpetka2001** [reply]
@catdadcode I just saw your edited comment with the new additions that you made. Of course, this is just a discussion and I by no chance am making decisions. Just make your case to the maintainer and he will be the one deciding. Exceptions can be made, if the maintainer thinks it's worth it. That's why they're called exceptions and I agree with you on that point as an exception.

**Comment by DrummyFloyd** [comment]
i'm agree with both point of view, so to not lose this, i'll move this discussion to `recipes`
